x-pack/plugins/apm/server/lib/services/get_service_transaction_groups_agg_results.ts

```diff
@@ -0,0 +1,182 @@
+  const { apmEventClient, start, end, esFilter } = setup;
+  const deltaAsMinutes = (end - start) / 1000 / 60;
+
+  const { intervalString } = getBucketSize({ start, end, numBuckets });
+
+  const field = getTransactionDurationFieldForAggregatedTransactions(
+    searchAggregatedTransactions
+  );
+
+  const response = await apmEventClient.search({
+    apm: {
+      events: [
+        getProcessorEventForAggregatedTransactions(
+          searchAggregatedTransactions
+        ),
+      ],
+    },
+    body: {
+      size: 0,
+      query: {
+        bool: {
+          filter: [
+            { term: { [SERVICE_NAME]: serviceName } },
+            { term: { [TRANSACTION_TYPE]: transactionType } },
+            { range: rangeFilter(start, end) },
+            ...getDocumentTypeFilterForAggregatedTransactions(
+              searchAggregatedTransactions
+            ),
+            ...esFilter,
+          ],
+        },
+      },
+      aggs: {
+        total_duration: { sum: { field: TRANSACTION_DURATION } },
+        transaction_name_filter: {
+          filter: { terms: { [TRANSACTION_NAME]: transactionNames } },
+          aggs: {
+            transaction_groups: {
+              terms: {
+                field: TRANSACTION_NAME,
+                size: transactionNames.length,
+              },
+              aggs: {
+                transaction_group_total_duration: {
+                  sum: { field: TRANSACTION_DURATION },
+                },
+                timeseries: {
+                  date_histogram: {
+                    field: '@timestamp',
+                    fixed_interval: intervalString,
+                    min_doc_count: 0,
+                    extended_bounds: {
+                      min: start,
+                      max: end,
+                    },
+                  },
+                  aggs: {
+                    ...getLatencyAggregation(latencyAggregationType, field),
+                    [EVENT_OUTCOME]: {
+                      filter: {
+                        term: { [EVENT_OUTCOME]: EventOutcome.failure },
+                      },
+                    },
+                  },
+                },
+              },
+            },
+          },
+        },
+      },
+    },
+  });
+
+  const buckets =
+    response.aggregations?.transaction_name_filter.transaction_groups.buckets ??
+    [];
+
+  const totalDuration = response.aggregations?.total_duration.value;
+
+  return buckets.reduce((acc, bucket) => {
+    const transactionName = bucket.key;
+
+    const latency: Coordinate[] = bucket.timeseries.buckets.map(
+      (timeseriesBucket) => ({
+        x: timeseriesBucket.key,
+        y: getLatencyValue({
+          latencyAggregationType,
+          aggregation: timeseriesBucket.latency,
+        }),
+      })
+    );
+
+    const throughput: Coordinate[] = bucket.timeseries.buckets.map(
+      (timeseriesBucket) => ({
+        x: timeseriesBucket.key,
+        y: timeseriesBucket.doc_count / deltaAsMinutes,
```

Comment:
This should be divided by bucketSize, rather than the entire time range (fwiw, the user doesn't see the difference because we don't show values in the spark plots).

Reply:
Really? because I was wondering to replace this calculation to the function @ogupte created to centralize the throughput calculation. https://github.com/elastic/kibana/blob/master/x-pack/plugins/apm/server/lib/helpers/calculate_throughput.ts#L9:L9

Reply:
Suppose that you have 400 transactions for 1 hour, then have 4 buckets each with 100 transactions. What happens when you calculate throughput by dividing the count per bucket by the entire time range, instead of the bucket size?

Reply:
FWICT dividing by the entire time range you have the buckets in transactions per minute and using the bucket size you have the buckets in percentage. Or am I missing something?

Reply:
This is where we could use the [rate agg](https://www.elastic.co/guide/en/elasticsearch/reference/7.x/search-aggregations-metrics-rate-aggregation.html) and avoid having to calculate ourselves, right?

Reply:
I've change to use the rate agg as suggested by @sqren .

The result is almost the same.
Left image (before) | Right image (after)
<img width="1906" alt="Screenshot 2021-02-04 at 13 06 19" src="https://user-images.githubusercontent.com/55978943/106891326-cea17400-66ea-11eb-854c-01b45163ce6c.png">

Reply:
@cauemarcondes do the values for the spark plot change (i.e., in the API response)?

Reply:
@dgieselaar yes, it changed, a few decimals.

Reply:
And what are you comparing to? dividing it by `bucketSize` or by `deltaAsMinutes`?

Reply:
`deltaAsMinutes`, I can compare dividing by bucketSize too

Reply:
If you have time, can you post the responses of:
- dividing it by (bucketSize * 60)
- dividing it by deltaAsMinutes
- using the rate agg

Reply:
@dgieselaar no problem.
- using the rate agg
```js
[
  {
    "x": 1607435820000,
    "y": 0,
  },
  {
    "x": 1607435880000,
    "y": 0.01,
  },
  {
    "x": 1607435940000,
    "y": 0.02,
  },
  {
    "x": 1607436000000,
    "y": 0.01,
  },
  {
    "x": 1607436060000,
    "y": 0,
  },
  {
    "x": 1607436120000,
    "y": 0.01,
  },
  {
    "x": 1607436180000,
    "y": 0,
  },
  {
    "x": 1607436240000,
    "y": 0.04,
  },
  {
    "x": 1607436300000,
    "y": 0,
  },
  {
    "x": 1607436360000,
    "y": 0,
  },
  {
    "x": 1607436420000,
    "y": 0,
  },
  {
    "x": 1607436480000,
    "y": 0.02,
  },
  {
    "x": 1607436540000,
    "y": 0,
  },
  {
    "x": 1607436600000,
    "y": 0.01,
  },
  {
    "x": 1607436660000,
    "y": 0,
  },
  {
    "x": 1607436720000,
    "y": 0,
  },
  {
    "x": 1607436780000,
    "y": 0,
  },
  {
    "x": 1607436840000,
    "y": 0,
  },
  {
    "x": 1607436900000,
    "y": 0,
  },
  {
    "x": 1607436960000,
    "y": 0.02,
  },
  {
    "x": 1607437020000,
    "y": 0,
  },
  {
    "x": 1607437080000,
    "y": 0,
  },
  {
    "x": 1607437140000,
    "y": 0,
  },
  {
    "x": 1607437200000,
    "y": 0.02,
  },
  {
    "x": 1607437260000,
    "y": 0,
  },
  {
    "x": 1607437320000,
    "y": 0,
  },
  {
    "x": 1607437380000,
    "y": 0,
  },
  {
    "x": 1607437440000,
    "y": 0,
  },
  {
    "x": 1607437500000,
    "y": 0,
  },
  {
    "x": 1607437560000,
    "y": 0,
  },
  {
    "x": 1607437620000,
    "y": 0,
  },
]
```

- dividing it by deltaAsMinutes
```js
[
  {
    "x": 1607435820000,
    "y": 0,
  },
  {
    "x": 1607435880000,
    "y": 0.0333333333333333,
  },
  {
    "x": 1607435940000,
    "y": 0.0666666666666667,
  },
  {
    "x": 1607436000000,
    "y": 0.0333333333333333,
  },
  {
    "x": 1607436060000,
    "y": 0,
  },
  {
    "x": 1607436120000,
    "y": 0.0333333333333333,
  },
  {
    "x": 1607436180000,
    "y": 0,
  },
  {
    "x": 1607436240000,
    "y": 0.133333333333333,
  },
  {
    "x": 1607436300000,
    "y": 0,
  },
  {
    "x": 1607436360000,
    "y": 0,
  },
  {
    "x": 1607436420000,
    "y": 0,
  },
  {
    "x": 1607436480000,
    "y": 0.0666666666666667,
  },
  {
    "x": 1607436540000,
    "y": 0,
  },
  {
    "x": 1607436600000,
    "y": 0.0333333333333333,
  },
  {
    "x": 1607436660000,
    "y": 0,
  },
  {
    "x": 1607436720000,
    "y": 0,
  },
  {
    "x": 1607436780000,
    "y": 0,
  },
  {
    "x": 1607436840000,
    "y": 0,
  },
  {
    "x": 1607436900000,
    "y": 0,
  },
  {
    "x": 1607436960000,
    "y": 0.0666666666666667,
  },
  {
    "x": 1607437020000,
    "y": 0,
  },
  {
    "x": 1607437080000,
    "y": 0,
  },
  {
    "x": 1607437140000,
    "y": 0,
  },
  {
    "x": 1607437200000,
    "y": 0.0666666666666667,
  },
  {
    "x": 1607437260000,
    "y": 0,
  },
  {
    "x": 1607437320000,
    "y": 0,
  },
  {
    "x": 1607437380000,
    "y": 0,
  },
  {
    "x": 1607437440000,
    "y": 0,
  },
  {
    "x": 1607437500000,
    "y": 0,
  },
  {
    "x": 1607437560000,
    "y": 0,
  },
  {
    "x": 1607437620000,
    "y": 0,
  },
]
```

- dividing it by (bucketSize * 60)
```js
[
  {
    "x": 1607435820000,
    "y": 0,
  },
  {
    "x": 1607435880000,
    "y": 0.00104166666666667,
  },
  {
    "x": 1607435940000,
    "y": 0.00208333333333333,
  },
  {
    "x": 1607436000000,
    "y": 0.00104166666666667,
  },
  {
    "x": 1607436060000,
    "y": 0,
  },
  {
    "x": 1607436120000,
    "y": 0.00104166666666667,
  },
  {
    "x": 1607436180000,
    "y": 0,
  },
  {
    "x": 1607436240000,
    "y": 0.00416666666666667,
  },
  {
    "x": 1607436300000,
    "y": 0,
  },
  {
    "x": 1607436360000,
    "y": 0,
  },
  {
    "x": 1607436420000,
    "y": 0,
  },
  {
    "x": 1607436480000,
    "y": 0.00208333333333333,
  },
  {
    "x": 1607436540000,
    "y": 0,
  },
  {
    "x": 1607436600000,
    "y": 0.00104166666666667,
  },
  {
    "x": 1607436660000,
    "y": 0,
  },
  {
    "x": 1607436720000,
    "y": 0,
  },
  {
    "x": 1607436780000,
    "y": 0,
  },
  {
    "x": 1607436840000,
    "y": 0,
  },
  {
    "x": 1607436900000,
    "y": 0,
  },
  {
    "x": 1607436960000,
    "y": 0.00208333333333333,
  },
  {
    "x": 1607437020000,
    "y": 0,
  },
  {
    "x": 1607437080000,
    "y": 0,
  },
  {
    "x": 1607437140000,
    "y": 0,
  },
  {
    "x": 1607437200000,
    "y": 0.00208333333333333,
  },
  {
    "x": 1607437260000,
    "y": 0,
  },
  {
    "x": 1607437320000,
    "y": 0,
  },
  {
    "x": 1607437380000,
    "y": 0,
  },
  {
    "x": 1607437440000,
    "y": 0,
  },
  {
    "x": 1607437500000,
    "y": 0,
  },
  {
    "x": 1607437560000,
    "y": 0,
  },
  {
    "x": 1607437620000,
    "y": 0,
  },
]
```

Reply:
@dgieselaar what do you think about the changes? do you think it is secure to move with rate agg?

Reply:
I should have said bucketSize/60, my bad. How does the rate aggregation look (the input) in your query?

Reply:
```js
throughput_rate: {
  rate: {
    field: TRANSACTION_DURATION,
    unit: 'minute',
    mode: 'value_count',
  },
},
```